Introduce FitAddon tests

leomoty opened the pull request:
Fixes #2167

I have a few questions:

- Can I rely on the default #terminal-container size?
- Is changing width good enough? should cover height as well?
- I noticed that if I try something stupid, like having a 1x1 pixel #terminal-container the fit will propose a negative cols. Is this expected behavior?

I want to remove the duplicated code to a helper function before merging.

Tyriar commented on addons/xterm-addon-fit/src/FitAddon.api.ts:
```diff
@@ -0,0 +1,100 @@
+    await page.evaluate(`
+      window.fit = new FitAddon();
+      window.term.loadAddon(window.fit);
+      document.querySelector('#terminal-container').style.width='100%';
```
This might be better as an absolute `px` value in case `#terminal-container`'s container changes its size?

leomoty replied:
Sounds good

Tyriar commented on addons/xterm-addon-fit/src/FitAddon.api.ts:
```diff
@@ -0,0 +1,100 @@
+    assert.equal(await page.evaluate(`window.fit.proposeDimensions()`), undefined);
+  });
+
+  it('default propose fit', async function(): Promise<any> {
```
Can you move all tests relating to just proposeDimensions into `describe('proposeDimensions', () => {`, and all fit ones into the same but `'fit'`?

leomoty replied:
Leave the `no terminal` test outside?

Tyriar replied:
The no terminal test just uses proposeDimensions so it can live in there

Tyriar commented on addons/xterm-addon-fit/src/FitAddon.api.ts:
```diff
@@ -0,0 +1,100 @@
+    await page.evaluate(`
+      window.fit = new FitAddon();
+      window.term.loadAddon(window.fit);
+      document.querySelector('#terminal-container').style.width='100%';
```
Again it might be safer to use `px`

leomoty commented:
How about this instead @Tyriar? it removes a bit of duplicated code, but we don't rely directly on the default width? I can revert otherwise.

I still need to add one test for the 1x1 case, will do after work.

leomoty commented:
I don't understand why the MacOS integration tests failed, I kept the same timeout that exists elsewhere?

Also, regarding the FitAddon, it doesn't need to be smaller than 1x1 to break:

```javascript
    it('small width', async function(): Promise<any> {
      await openTerminal();
      await loadFit(1);
      console.log(await page.evaluate(`window.fit.proposeDimensions()`));
    });
```
returns:

```javascript
{ cols: -2, rows: 26 }
```

What is the desired result? 

```javascript
    it('small width', async function(): Promise<any> {
      await openTerminal();
      await loadFit(1);
      await page.evaluate(`window.fit.fit()`);
      console.log(await page.evaluate(`window.term.cols`));
      console.log(await page.evaluate(`window.term.rows`));
    });
```
returns:
```
2
26
```

leomoty commented:
Checking the `BufferService.ts`, I would expect FitAddon to match MINIMUM_COLS and MINIMUM_ROWS defined there. Since that is what Terminal seems to do nonetheless.

Tyriar commented:
> I don't understand why the MacOS integration tests failed, I kept the same timeout that exists elsewhere?

Not sure, try narrow down which await is not returning? I'll requeue the build in case it's flakiness

leomoty commented:
I have updated the `loadFit` to allow both width/height (defaults are based from demo's style.css).

Also updated FitAddon to respect MINIMUM_COLS and MINIMUM_ROWS.

Mac OS is still flaky, but since I only have linux to test, I am not sure why.

Tyriar commented:
Looks good, I'll look into the mac test now

Tyriar commented:
I couldn't repro the flakiness on my mac, I pushed a change to avoid loading the page every time though which makes the tests take about 25% of the time. Hopefully the flakiness is fixed magically 🤞
